src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -283,43 +294,64 @@ internal string GetHeaderString(HeaderDescriptor descriptor)
 
                 // Note that if we get multiple values for a header that doesn't support multiple values, we'll
                 // just separate the values using a comma (default separator).
-                string? separator = descriptor.Parser != null && descriptor.Parser.SupportsMultipleValues ? descriptor.Parser.Separator : HttpHeaderParser.DefaultSeparator;
-                return string.Join(separator, multiValue!);
+                return string.Join(descriptor.Separator, multiValue!);
             }
 
             return string.Empty;
         }
 
         #region IEnumerable<KeyValuePair<string, IEnumerable<string>>> Members
 
-        public IEnumerator<KeyValuePair<string, IEnumerable<string>>> GetEnumerator() => _headerStore != null && _headerStore.Count > 0 ?
-                GetEnumeratorCore() :
-                ((IEnumerable<KeyValuePair<string, IEnumerable<string>>>)Array.Empty<KeyValuePair<string, IEnumerable<string>>>()).GetEnumerator();
+        public IEnumerator<KeyValuePair<string, IEnumerable<string>>> GetEnumerator() => _headerStore.IsEmpty ?
+                ((IEnumerable<KeyValuePair<string, IEnumerable<string>>>)Array.Empty<KeyValuePair<string, IEnumerable<string>>>()).GetEnumerator() :
+                GetEnumeratorCore();
```

Comment:
Why not to fold `GetEnumeratorCore` here?

Reply:
It was done so that calling `GetEnumerator` on an empty header collection may return an object-backed `IEnumerator`, thus avoiding an allocation.

I'm not sure it's the most critical optimization ever since enumerating empty headers should be rare.

Reply:
What I meant was to have line 
```
if (_headerStore.IsEmpty) return Array.Empy<KeyValuePair<string, IEnumerable<string>>()`
```
inside `GetEnumeratorCore`

Reply:
That would mean that you are forced to allocate the `GetEnumeratorCore` state machine even when empty.

Reply:
ok, missed the yield at the last line